### PR TITLE
operator: validate kube-controller-manager config before starting deployment of a new revision

### PR DESCRIPTION
**Pull request by sttts**

### CODE REVIEW

**Comment by deads2k** [comment]
Rather than do this, can we have a graceful shutdown that updates our lease to a super short time before death?

/hold

**Comment by deads2k** [comment]
When you rebase, please target specific items in the config and provide an event and open an issue to mark us as failing or unavailable or something.

/hold cancel

**Comment by sttts** [comment]
/hold

addressing https://github.com/openshift/cluster-kube-controller-manager-operator/pull/96#issuecomment-443933270 first.

File: pkg/operator/target_config_reconciler_v311_00.go [hunk 1]
```diff
@@ -82,6 +85,14 @@ func createTargetConfigReconciler_v311_00_to_latest(c TargetConfigReconciler, re
 }
 
 func manageKubeControllerManagerConfigMap_v311_00_to_latest(client coreclientv1.ConfigMapsGetter, recorder events.Recorder, operatorConfig *v1alpha1.KubeControllerManagerOperatorConfig) (*corev1.ConfigMap, bool, error) {
+	var observedConfig map[string]interface{}
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/operator/target_config_reconciler_v311_00.go]: I would keep this logic here and set the condition for the installer controller and block installer controller probably if the condition is true? (that will prevent running installation of revision we know already that will fail)

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/operator/target_config_reconciler_v311_00.go]: @deads2k need your opinion here

File: pkg/operator/configobservation/validation/validate.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package validation
+
+import "github.com/openshift/cluster-kube-controller-manager-operator/pkg/operator/configobservation/network"
+
+// Validate verifies whether the observed configuration can be used to make progress.
+func Validate(observedConfig map[string]interface{}) (errs []error) {
```
Comment (deads2k) [inline pkg/operator/configobservation/validation/validate.go]: this is correct, but extremely broad for abuse and drift.  How about a simple list of required keys?

File: vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/revision/deployment_controller.go [hunk 3]
```diff
@@ -106,6 +115,22 @@ func (c RevisionController) createRevisionIfNeeded(operatorSpec *operatorv1.Oper
 		return true, nil
 	}
 
+	if c.validate != nil {
+		if err := c.validate(configs, secrets); err != nil {
+			cond := operatorv1.OperatorCondition{
+				Type:    "RevisionControllerFailing",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "ContentInvalid",
+				Message: err.Error(),
```
Comment (deads2k) [inline vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/revision/deployment_controller.go]: that helper function I described to @mfojtik here seems to apply https://github.com/openshift/library-go/pull/150#discussion_r244737541

File: vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/revision/deployment_controller.go [hunk 4]
```diff
@@ -106,6 +115,22 @@ func (c RevisionController) createRevisionIfNeeded(operatorSpec *operatorv1.Oper
 		return true, nil
 	}
 
+	if c.validate != nil {
+		if err := c.validate(configs, secrets); err != nil {
```
Comment (deads2k) [inline vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/revision/deployment_controller.go]: doesn't this validation need to happen *before* we create the configmaps and secrets for the revision?

Comment (sttts) [reply vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/revision/deployment_controller.go]: Does not matter much. We have to make sure the config synched to the revisioned variants is correct, all at once and atomically checked. We could read all configmaps and secrets at once, validate them and sync those. But we cannot (form atomicity reasons) validate by reading, then read again to sync. Hence, my change is small, maybe slightly ugly. Valdating before the sync means to change the whe sync logic. Is it worth it?

File: vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/revision/deployment_controller.go (hunk 3 above)
Comment (sttts) [reply vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/revision/deployment_controller.go]: fixed to use `NewMultiLineAggregate`

**Pull request closed without merging**